gufe/protocols/protocolunit.py

```diff
@@ -7,79 +7,312 @@
 """
 
 import abc
+from dataclasses import dataclass
+import traceback
+import uuid
 from os import PathLike
+from copy import copy
 from typing import Iterable, List, Dict, Any, Optional, Union
+import tempfile
 
-from .results import (
-    ProtocolUnitResult, ProtocolUnitFailure,
-)
+from ..tokenization import GufeTokenizable, GufeKey, normalize
 
 
-class ProtocolUnit(abc.ABC):
-    """A unit of work within a `ProtocolDAG`"""
+@dataclass
+class Context:
+    """Data class for passing around execution context components to
+    `ProtocolUnit._execute`.
+
+    """
+    unit_scratch: PathLike
+    dag_scratch: PathLike
+
+
+class ProtocolUnitResultBase(GufeTokenizable):
+    inputs: Dict[str, Any]   # `ProtocolUnitResult`s this `ProtocolUnitResult` is dependent on
+
+    name: Optional[str]      # name of the `ProtocolUnit` that produced this `ProtocolUnitResult`
+    pure: bool               # whether `ProtocolUnit` that produced this `ProtocolUnitResult` was a function purely of its inputs
+    source_key: GufeKey             # key of the `ProtocolUnit` that produced this `ProtocolUnitResult`
+
+    outputs: Dict[str, Any]  # outputs is a dict returned by a `ProtocolUnit`'s `_execute` method
```

Comment:
Maybe instead of Any we can actually have a type which has the accepted types? Might catch some snafus with static typing

Reply:
Can you show me how to do this? My `typing` foo is still a bit limited. Would it be something like:

```
GufeAcceptable = Union[Dict[str,GufeAcceptable],List[GufeAcceptable],int,float,bool,...]
```
?